### PR TITLE
Use local private credentials (json key file) to refresh GCS access token

### PR DESCRIPTION
**Pull request by beinan**
Generating and refresh GCS access token by the local private credentials (json key file)
Support the credentials of either "service_account" or "authorized_user" type
Added to both presto-cli and presto-jdbc

```
== RELEASE NOTES ==

General Changes
* Use local private credentials (json key file) to refresh GCS access token
presto-cli --extra-credential hive.gcs.credentials.path="${PRIVATE_KEY_JSON_PATH}"
```

### CODE REVIEW

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        catch (IOException e) {
+            throw new ClientException("Google credential refreshing error", e);
+        }
+        return credentials;
+    }
+
+    private GoogleCredentials createCredentials()
+    {
+        try {
+            return GoogleCredentials.fromStream(new FileInputStream(credentialsFilePath)).createScoped(Collections.singleton("https://www.googleapis.com/auth/devstorage.read_only"));
```
Comment (ChunxuTang) [inline presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:76]: Do we need to hardcode the URL here? Maybe we can move it to a static variable.

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    @Override
+    public Response intercept(Chain chain)
+            throws IOException
+    {
+        return chain.proceed(attachGCSAccessToken(chain.request()));
+    }
+
+    private Request attachGCSAccessToken(Request request)
+    {
+        AccessToken token = getCredentials().getAccessToken();
```
Comment (ChunxuTang) [inline presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:53]: The GCP documentation here recommends creating a signed JWT with the client ID and a private key from the GCP credential. Is this step automatically covered in the Google OAuth libs?

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        return request.newBuilder()
+                .addHeader(PRESTO_EXTRA_CREDENTIAL, GCS_CREDENTIALS_OAUTH_TOKEN_KEY + "=" + token.getTokenValue())
+                .build();
+    }
+
+    private synchronized GoogleCredentials getCredentials()
```
Comment (ChunxuTang) [inline presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:59]: Is the `synchronized` keyword necessary here? Will this function be invoked in a multi-threaded situation?

Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:59]: Good call.  This interceptor is used by presto-cli and presto-jdbc.  I think presto-jdbc might have some queries running in parallel.  Even for a single query, during the http conversation,  there are still multiple http calls, some of which might be triggered in an asynchronous way.

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java (hunk 2 above)
Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:53]: Really a good point, I guess not.

Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:53]: Good call, I guess not.  I will post another commit or PR to support JWT with a given client ID

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java (hunk 1 above)
Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:76]: Good catch,  will extract it as a constant val

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java (hunk 3 above)
Comment (ChunxuTang) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthInterceptor.java:59]: Gotcha. Thanks for your explanation.

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_EXTRA_CREDENTIAL;
+import static java.util.Objects.requireNonNull;
+
+public class GCSOAuthHandler
+        implements Interceptor, Authenticator
+{
+
+    public static final String GCS_CREDENTIALS_PATH_KEY = "hive.gcs.credentials.path";
+    private static final String GCS_CREDENTIALS_OAUTH_TOKEN_KEY = "hive.gcs.oauth";
+    private final ClientSession session;
```
Comment (zhenxiao) [inline presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: add one blank line between private static final and private final

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.Collections;
+
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_EXTRA_CREDENTIAL;
+import static java.util.Objects.requireNonNull;
+
+public class GCSOAuthHandler
+        implements Interceptor, Authenticator
+{
+
+    public static final String GCS_CREDENTIALS_PATH_KEY = "hive.gcs.credentials.path";
```
Comment (zhenxiao) [inline presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: shall we keep hive here? kind of misleading. If the session property is already defined, could we import the string from other class?

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    private static final String GCS_CREDENTIALS_OAUTH_TOKEN_KEY = "hive.gcs.oauth";
+    private final ClientSession session;
+
+    private String credentialsFilePath;
+
```
Comment (zhenxiao) [inline presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: remove this blank line

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        catch (IOException e) {
+            throw new ClientException("Google credential refreshing error", e);
+        }
+        return credentials;
+    }
+
+    private GoogleCredentials createCredentials()
+    {
+        try {
+            return GoogleCredentials.fromStream(new FileInputStream(credentialsFilePath)).createScoped(Collections.singleton("https://www.googleapis.com/auth/devstorage.read_only"));
```
Comment (zhenxiao) [inline presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: make constant:
https://www.googleapis.com/auth/devstorage.read_only

File: pom.xml [hunk 8]
```diff
@@ -548,7 +548,7 @@
             <dependency>
                 <groupId>com.facebook.presto.hadoop</groupId>
                 <artifactId>hadoop-apache2</artifactId>
-                <version>2.7.4-7</version>
+                <version>2.7.4-8-beinan</version>
```
Comment (zhenxiao) [inline pom.xml]: oh, this PR is a dependency of:
https://github.com/prestodb/presto-hadoop-apache2/pull/43

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriverUri.java [hunk 9]
```diff
@@ -186,6 +188,10 @@ public void setupClient(OkHttpClient.Builder builder)
                                 .orElseGet(() -> defaultCredentialCachePath().map(File::new).orElse(null))));
             }
 
+            Map<String, String> extraCredentials = EXTRA_CREDENTIALS.getValue(properties).orElse(ImmutableMap.of());
+            Optional.ofNullable(extraCredentials.get(GCSOAuthInterceptor.GCS_CREDENTIALS_PATH_KEY))
```
Comment (zhenxiao) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriverUri.java:192]: static import GCSOAuthInterceptor.GCS_CREDENTIALS_PATH_KEY

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java (hunk 5 above)
Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: The "hive.gcs.credentials.path" is the key I newly added.   The other 'hive.gcs.oauth' is an existing one, which can be also found in presto-hive. 

presto-hive/src/main/java/com/facebook/presto/hive/gcs/GcsConfigurationProvider.java
28:    private static final String GCS_OAUTH_KEY = "hive.gcs.oauth";

If we imported the constant "hive.gcs.oauth" from presto-hive,  we might introduce a new dependency from presto-client to presto-hive.  Shall we do that?

I'm just thinking the name of these two keys need to be consistent.

Comment (zhenxiao) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: get it. no need to have dependency from presto-hive
keep `hive.gcs.credentials.path` is fine

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriverUri.java (hunk 9 above)
Comment (beinan) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriverUri.java]: done

File: pom.xml (hunk 8 above)
Comment (beinan) [reply pom.xml]: We have updated the version to 2.7.4-8 in another pr

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java (hunk 7 above)
Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: done

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java (hunk 6 above)
Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: done

File: presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java (hunk 4 above)
Comment (beinan) [reply presto-client/src/main/java/com/facebook/presto/client/GCSOAuthHandler.java]: done

File: pom.xml [hunk 10]
```diff
@@ -1174,7 +1185,6 @@
                     </exclusion>
                 </exclusions>
             </dependency>
-
```
Comment (zhenxiao) [inline pom.xml:1177]: shall we keep the blank line?

File: pom.xml [hunk 11]
```diff
@@ -1158,7 +1170,6 @@
...
```
Comment (zhenxiao) [inline pom.xml:1161]: keep the blank line for format

Comment (beinan) [reply pom.xml]: done

File: pom.xml (hunk 10 above)
Comment (beinan) [reply pom.xml]: done

**Pull request merged**